src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java

```diff
@@ -75,4 +103,56 @@ public OneTimePasswordResource(OneTimePasswordService oneTimePasswordService) {
         oneTimePasswordService.check(oneTimePasswordCheckDto);
         return ResponseEntity.ok(oneTimePasswordCheckDto);
     }
+
+    /**
+     * POST  /one-time-password/validate : Validate an existing oneTimePassword and redirect page.
+     * @see OneTimePasswordResource#checkOneTimePassword(OneTimePasswordCheckDto)
+     */
+    @PreAuthorize("hasPermission({'oneTimePasswordCheckDto': #oneTimePasswordCheckDto} ,'OTP.ONETIMEPASSWORD.CHECK')")
+    @PostMapping("/one-time-password/validate")
+    @Timed
+    public RedirectView checkOneTimePasswordAndRedirectWithCode(@Valid @RequestBody OneTimePasswordCheckDto oneTimePasswordCheckDto,
+                                                                @RequestHeader(value = "redirect-uri") String redirectUri) throws URISyntaxException {
+        redirectUri = decodeUrl(redirectUri);
+        checkOneTimePassword(oneTimePasswordCheckDto);
+        OneTimePasswordDto dto = oneTimePasswordService.findOne(oneTimePasswordCheckDto.getId()).get();
+        String code =  uaaRepository.getOAuth2Token(ImmutableMap.of(MSISDN, dto.getReceiver()));
+        return new RedirectView(new StringBuilder(redirectUri).append("?code=").append(code).toString(), true);
+    }
+
+    /**
+     *
+     * POST /oauth/token  Convert  "code" parameter into "access_token" in order to satisfy OAuth protocol
+     * @param code jwt token
+     */
+    @PostMapping("/oauth/token")
+    @Timed
+    public ResponseEntity validateCode(@RequestParam(name = "code") String code) {
+        return ResponseEntity.ok(ImmutableMap.of(ACCESS_TOKEN, code));
+    }
+
+    /**
+     * GET /userinfo  Return user information gotten from additional details of access_token
+     */
+    @GetMapping("/userinfo")
+    @Timed
+    public ResponseEntity getUserInfo() {
+        Optional<String> login = authenticationContextHolder.getContext().getAdditionalDetailsValue(MSISDN);
+        if (!login.isPresent()) {
+            log.warn("MSISDN is not set in access token");
+            return ResponseEntity.badRequest().build();
+        }
+        String phoneNumber = login.get();
+        return ResponseEntity.ok(new HashMap<String, String>(){{
```

Comment:
maybe better to user ImmutableMap.of
and static import?
of("phoneNumber", phoneNumber, "id", phoneNumber)